sql/src/main/java/io/crate/protocols/postgres/BulkPortal.java

```diff
@@ -136,14 +134,13 @@ public void execute(ResultReceiver resultReceiver, int maxRows) {
         return executeBulk(portalContext.getExecutor(), plan, jobId, statsTables);
     }
 
-    private ListenableFuture<Void> executeBulk(Executor executor, Plan plan, final UUID jobId,
-                             final StatsTables statsTables) {
-        final SettableFuture<Void> future = SettableFuture.create();
-        List<? extends ListenableFuture<Long>> futures = executor.executeBulk(plan);
-        ListenableFuture<List<Long>> allAsList = Futures.allAsList(futures);
-        Futures.addCallback(allAsList, new FutureCallback<List<Long>>() {
-                @Override
-                public void onSuccess(@Nullable List<Long> result) {
+    private CompletableFuture<Void> executeBulk(Executor executor, Plan plan, final UUID jobId,
+                                                final StatsTables statsTables) {
+        final CompletableFuture<Void> future = new CompletableFuture<>();
+        List<CompletableFuture<Long>> futures = executor.executeBulk(plan);
+        CompletableFuture<List<Long>> allFuturesCollected = CompletableFutures.allSuccessfulOrFailedFuture(futures);
```

Comment:
That method has the advantage of collecting the results in order and failing fast (on first future failure). I think using a callback per future, all manipulating the resultReceivers, will make this piece of code messier (synchronization or a thread safe data structure will be required, and also the order of the results will have to be maintained externally - like an index passed as an argument to the future.whenComplete consumer or something like that)